File: src/main/java/ravex/modules/render/FreeLook.java
package ravex.modules.render;

import ravex.modules.Category;
import ravex.modules.Module;
import net.minecraft.client.Minecraft;

/**
 * FreeLook Module
 * Decouples camera rotation in third-person view with wall-collision protections.
 */
public class FreeLook extends Module {
    public static final FreeLook INSTANCE = new FreeLook();

    private float lookYaw = 0.0f;
    private float lookPitch = 0.0f;
    private int originalPerspective = 0; // 0 = FIRST_PERSON

    private FreeLook() {
        super("FreeLook", Category.RENDER);
    }

    @Override
    protected void onEnable() {
        Minecraft mc = Minecraft.getInstance();
        if (mc.player != null) {
            lookYaw = mc.player.getYRot();
            lookPitch = mc.player.getXRot();
            originalPerspective = mc.options.getCameraType().ordinal();
            // Force 3rd person back perspective
            mc.options.setCameraType(net.minecraft.client.CameraType.THIRD_PERSON_BACK);
        }
    }

    @Override
    protected void onDisable() {
        Minecraft mc = Minecraft.getInstance();
        if (mc.options != null) {
            // Restore perspective
            var types = net.minecraft.client.CameraType.values();
            if (originalPerspective >= 0 && originalPerspective < types.length) {
                mc.options.setCameraType(types[originalPerspective]);
            }
        }
    }

    public void turn(double yRot, double xRot) {
        // Direct, crisp looking controls matching vanilla mouse handler feeling exactly
        lookYaw += (float) yRot;
        lookPitch += (float) xRot;
        
        // Clamp pitch to avoid turning upside down
        lookPitch = Math.max(-90.0f, Math.min(90.0f, lookPitch));
    }

    public float getLookYaw() {
        return lookYaw;
    }

    public float getLookPitch() {
        return lookPitch;
    }
}

File: src/main/kotlin/volthack/modules/render/FreeLook.kt
package volthack.modules.render

import net.minecraft.client.CameraType
import net.minecraft.client.Minecraft
import volthack.setting.Category
import volthack.setting.Module

object FreeLook : Module("FreeLook", "Free camera in third-person without changing player movement", Category.RENDER) {
    private val speed by float("Speed", 0.6f, 0.1f, 2.0f, 0.05f)

    var freeYaw = 0f
    var freePitch = 0f
    private var prevCameraType: CameraType? = null

    override fun onEnable() {
        val mc = Minecraft.getInstance()
        val player = mc.player ?: return
        freeYaw = player.yRot
        freePitch = player.xRot
        prevCameraType = mc.options.cameraType
        if (mc.options.cameraType == CameraType.FIRST_PERSON) {
            mc.options.cameraType = CameraType.THIRD_PERSON_BACK
        }
    }

    override fun onDisable() {
        val mc = Minecraft.getInstance()
        prevCameraType?.let { mc.options.cameraType = it }
        prevCameraType = null
    }

    fun onMouseTurn(yRot: Double, xRot: Double) {
        freeYaw += (yRot * speed).toFloat()
        freePitch += (xRot * speed).toFloat()
        freePitch = freePitch.coerceIn(-90f, 90f)
    }
}
